### ISSUE TITLE
Enhancing the connection pool

### ISSUE THREAD

**Issue by Riron** [prisma/prisma#19215]
## Problem

Those problems are not new and have been [mentioned before](https://github.com/prisma/prisma/issues/6156#issuecomment-1034051180): 
1. Theidle lifetime is only lazily evaluated.

Under normal load we have around 400 connections to our DB. Several times per day however we have spikes that increase the number of connection to our 800 limit. Because the mobc cleaner never kicks in, our database stays at 800 connections forever. With half of them doing nothing four hours and hours. It consumes a lot of unnecessary RAM, and can sometimes cause our psql node to swap and eventually be killed by the OS. To prevent swaping we have to manually kill the connections that have been idle for too long . A [PR has been open for 2 years on mobc](https://github.com/importcjj/mobc/pull/60) about that.

2. Connection health is not checked on every take from the pool.

When one of our Psql node is killed, a backup node kicks in but all the connections that were connected to the master node are closed. However mobc still returns those connection to the Prisma backend, causing the request to fail. When you have 800 connections, this could mean 800 failing requests.

It seems that the Prisma team is spending a lot of time optimizing for serverless context where such issues probably don't really matter. But I feel like the connection pool is one of Prisma's weak point and causes many problems for big non serverless apps.

## Suggested solution

Either enhance or rewrite of the connection pool 😄 
I don't want to throw mobc under the bus by the way, it's a great tool ! I just feel under certain scenarios it has a few lacking features.

## Alternatives

I've [been experimenting](https://github.com/Riron/swim) with importcjj [idea of a lock free connection pool](https://github.com/importcjj/mobc/pull/60#issuecomment-879616110) using a single threaded event loop and it seems to have significantly better performance. But I am no expert in connection pool and maybe tools like r2d2 or deadpool could be alternatives to mobc ?

Maybe Prisma could give users the possibility to use different connection pool through preview features. That would allow to use the right tool for the right job.

**Comment by Riron** [comment]
I know the team has a busy agenda, but I just want to add that if you need help moving forward on that issue I would be willing to contribute !

**Comment by janpio** [comment]
> 3\. Connection health is not checked on every take from the pool.
> 
> When one of our Psql node is killed, a backup node kicks in but all the connections that were connected to the master node are closed. However mobc still returns those connection to the Prisma backend, causing the request to fail. When you have 800 connections, this could mean 800 failing requests.

Do you have a suggestion how we could deal with that problem?

**Comment by Riron** [comment]
The manager exposes two kind of checks:
- [fast & synchronous](https://github.com/importcjj/mobc/blob/main/src/lib.rs#L151) - typically, is the TCP connection still open ? eg I am guessing its what the `is_closed` [methods checks for in tokio_postgres](https://docs.rs/tokio-postgres/latest/tokio_postgres/struct.Client.html#method.is_closed) but I haven't checked
- [slow & async](https://github.com/importcjj/mobc/blob/main/src/lib.rs#L147) - typically with SQL, `SELECT 1;`

I think the first kind of check should always be called before returning the connection to the client. And the second kind of check should be optional. Meaning you would have a *good enough* check by default, and could have *guarantees* if you are willing to pay a latency tax.

Mobc does the "quick check" only when [it recycles a connection](https://github.com/importcjj/mobc/blob/main/src/lib.rs#L590).

[In my experiment](https://github.com/Riron/swim/blob/main/src/lib.rs#LL218C1-L218C1) I always do a quick check before returning a connection. And the slow check can be activated through config.

**Comment by jkomyno** [comment]
Some additional notes on this: @Sytten suggested us to take a look at https://github.com/bikeshedder/deadpool as a potential replacement of `mobc`, to be used in our Rust Prisma engines.

**Comment by Sytten** [comment]
@jkomyno Yeah mobc would need a full rewrite at this point, I talked with the author and he started it in another repo (https://github.com/importcjj/mobc2-prototype) but didn't finish it. I would highly suggest looking at deadpool and engaging with the maintainer (he is very active). You can likely give him a small contract work like I did in the past for prisma to build stuff that is missing.
I remember talking with some folks at prisma at the time that wanted a lock-free pool like the sqlx one (deadpool uses a mutex), but then I talked with the sqlx folks who said their design wasn't particularly great either 😅
Personally I think a well placed lock is fine and that wont be the blocker for performances. 

I agree with @Riron on the two big problems of mobc, but for point 2 checking the health of the connection on each checkout is costly (in terms of latency). In theory Prisma (Quaint) already knows when most TCPStreams die, but currently that info doesn't trickle back up to the pool. That would already be a huge improvement. Sometimes though the stream is only partially closed so you don't get a broken pipe until you write to it, that requires a change in the engine (as opposed to quaint for the other items) where it would retry a certain number of times on certain conditions (even better if the user can configure it). That would solve 99.9% of the problems for normal usage, though there are some edge cases where it might be problematic (mostly unreliable networking problems, which I used to get quite often on GCP Cloud Run).
